nl/nl.1
'\" t
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)nl.1	1.10 (gritter) 8/14/05
.TH NL 1 "8/14/05" "Heirloom Toolchest" "User Commands"
.SH NAME
nl \- line numbering filter
.SH SYNOPSIS
.HP
.ad l
\fBnl\fR
[\fB\-h\fItype\fR]
[\fB\-b\fItype\fR]
[\fB\-f\fItype\fR]
[\fB\-v\fIstart#\fR]
[\fB\-i\fIincr\fR]
[\fB\-p\fR]
[\fB\-l\fInum\fR]
[\fB\-w\fIwidth\fR]
[\fB\-n\fIformat\fR]
[\fB\-d\fIdelim\fR]
[\fIfile\fR]
.br
.ad b
.SH DESCRIPTION
The
.B nl
command
prints a
.IR file ,
or, when no
.I file
argument is given,
standard input,
preceding lines with line numbers on the left.
.PP
Lines with special content are taken as page delimiters.
Three parts of pages are distinguished:
header, body and footer.
Each part is started when a line consisting only of the
following characters appears in input:
.RS
.sp
.TS
l l
lfB l.
Line contents	Start of
\e:\e:\e:	header
\e:\e:	body
\e:	footer
.TE
.sp
.RE
Line numbers are reset each time a header begins.
The start of the file is considered a body part
until a line marking another part appears.
.PP
The following options alter the behavior of
.BR nl :
.TP
.BI \-b type
Specifies which lines in a body part are numbered, where
.I type
is one of the following:
.RS
.TP
.B a
Number all lines,
unless otherwise specified by the
.B \-l
option.
.TP
.B t
Number all non-empty lines.
.TP
.B n
Number no lines.
.TP
.BI p string
Number all lines that match the regular expression
.IR string .
Regular expressions are as in
.IR ed (1);
.B /usr/5bin/nl
uses simple regular expressions,
.BR /usr/5bin/posix/nl ,
.BR /usr/5bin/posix2001/nl ,
and
.B /usr/5bin/s42/nl
use basic regular expressions.
.RE
.IP
The default is
.BR t .
.TP
.BI \-d delim
Change the special characters that delimit parts of a page
to the two characters in
.IR delim .
If
.I delim
consists of only one character,
the second character remains the default, `\fB:\fR'.
.TP
.BI \-f type
Specifies which lines in a footer part are numbered,
where
.I type
has the format described for
.BR \-b .
The default is
.BR n .
.TP
.BI \-h type
Specifies which lines in a header part are numbered,
where
.I type
has the format described for
.BR \-b .
The default is
.BR n .
.TP
.BI \-i incr
Line numbers are increased by
.IR incr ,
which defaults to 1.
.TP
.BI \-l num
For the
.BR \-ba ,
.BR \-fa ,
and
.B \-ha
options,
specifies which line of adjacent blank lines is printed.
The default is 1 (print all blank lines).
.TP
.BI \-n format
Sets the format of line numbers, where
.I format
is one of:
.RS
.TP
.B ln
left-justified, preceded by blanks
.TP
.B rn
right-justified, preceded by blanks
.TP
.B rz
right-justified, preceded by zeros
.RE
.IP
The default is
.BR rn .
.TP
.B \-p
Line numbers are not reset at the start of each header
when this option is present.
.TP
.BI \-s sep
Sets the string that separates line number and line content in output,
default is a tabulator.
.TP
.BI \-v start#
Each page is numbered starting at
.IR start# ,
which defaults to 1.
.TP
.BI \-w width
Sets the width of line numbers.
If a number has fewer digits than
.IR width ,
it is padded as described for the
.B \-n
option;
if it has more digits,
only the last
.I width
digits are printed.
The default
.I width
is 6.
.PP
The
.I file
argument can appear between options.
For
.B /usr/5bin/nl
and
.BR /usr/5bin/s42/nl ,
option and option argument must not be separated;
if an option is not immediately followed by its argument,
the default value is used.
.B /usr/5bin/posix/nl
and
.B /usr/5bin/posix2001/nl
require an option argument
and accept it regardless of space between option and argument.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Affects the collation order for range expressions,
equivalence classes, and collation symbols in basic
regular expressions.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters.
.SH EXAMPLES
.PP
To number all lines in a file,
effectively disabling header and footer recognition
since no line can contain a second newline character
and thus the condition is never met:
.RS
.sp
.nf
nl \-ba \-d\'
\&\' file
.fi
.sp
.RE
Note that there is no white space
other than the newline character
between \' \'.
.PP
To number all lines that are not empty
and contain at least one non-blank character:
.RS
.sp
.nf
nl \-bp\'[^      ]\' file
.fi
.sp
.RE
The characters between [ ] are:
.\" -- not really, don't mess up troff --
a circumflex, a space, and a tabulator.
.SH "SEE ALSO"
ed(1),
pr(1),
locale(7)
.SH NOTES
Since the default width causes line numbers
to be truncated if they consist of more than six digits, the
.B \-w
option should be used to increase this value
if higher numbers are to be expected.


CHANGES
Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such
  as "%7$s", that specify which argument is used next.
* pg: If a line does not fit on the screen, do not print a newline to
  force wraparound unless it is really necessary. This allows to select
  words across the wrap for copy-and-paste in an xterm.
* pg: Discard the ESC[48m sequence on a terminal that supports colors.
* ed: Do not print "expecting `w'" with the "q" command after a zero-byte
  file has been written.
* man: The tmac.an macros no longer use "@" as an escape character; its
  previous use lead to weird output with Unix nroff when an "@" appeared
  in titles or hanging labels.
* man: The "TCAT" post processor is now invoked separately for each manual
  entry. Previously, when the -t option was given with multiple entries and
  TCAT=dpost, only the first entry was actually printed.
* Some issues with building on AIX have been fixed.

Release 050713
* The 'echo' and 'test' commands have been added as standalone versions of
  those of the Heirloom Bourne shell.
* The 'false', 'kill', 'printf', 'pwd', and 'true' commands have been added.
* ed: Copying a larger number of lines with the 't' command could cause a
  segmentation fault because some pointers were not updated after calling
  realloc().
* ed: Do not warn about an 'input error' anymore when reading a file that
  lacks the terminating newline.
* ed: Disallow the 'u' command if no change has been made yet after a file
  specified on the command line has been read.

nl/nl.c
/*
 * nl - line numbering filter
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, January 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#if defined (S42)
static const char sccsid[] USED = "@(#)nl_s42.sl	1.18 (gritter) 5/29/05";
#elif defined (SU3)
static const char sccsid[] USED = "@(#)nl_su3.sl	1.18 (gritter) 5/29/05";
#elif defined (SUS)
static const char sccsid[] USED = "@(#)nl_sus.sl	1.18 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)nl.sl	1.18 (gritter) 5/29/05";
#endif

#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<wchar.h>
#include	<locale.h>

#if defined (SUS) || defined (SU3) || defined (S42)
#include	<regex.h>
#else
#include	<regexpr.h>
#endif
#include	<iblok.h>
#include	<mbtowi.h>

struct	type {
	const char	*t_str;
	int	t_chr;
	int	t_type;
	int	t_def;
#if defined (SUS) || defined (SU3) || defined (S42)
	regex_t	t_re;
#else
	char	*t_ep;
#endif
};

static struct type	btype = { "BODY", 'b', 't', 't' };
static struct type	ftype = { "FOOTER", 'f', 'n', 'n' };
static struct type	htype = { "HEADER", 'h', 'n', 'n' };

static wint_t		delim[2] = { '\\', ':' };	/* section delimiter */

static long		incr = 1;		/* line number increment */

static long		whichblank = 1;		/* blank line maximum */
static long		blanks;			/* blank line count */

static enum {
	FMT_LN,
	FMT_RN,
	FMT_RZ
} format = FMT_RN;

static int		pflag;			/* don't restart at pages */

static const char	*sep = "\t";		/* number/text separator */

static long		startnum = 1;		/* initial number */
static long		width = 6;		/* line number width */

static struct iblok	*input;			/* input file */

static unsigned long long	number;		/* line number */

static int		mb_cur_max;		/* MB_CUR_MAX acceleration */

/*
 * Get next character from string s and store it in wc; n is set to
 * the length of the corresponding byte sequence.
 */
#define	next(wc, s, n) (mb_cur_max > 1 && *(s) & 0200 ? \
			((n) = mbtowi(&(wc), (s), mb_cur_max), \
			 (n) = ((n) > 0 ? (n) : (n) < 0 ? (wc=WEOF, 1) : 1)) :\
		((wc) = *(s) & 0377, (n) = 1))

static void *
srealloc(void *vp, size_t nbytes)
{
	void	*p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, "input line too long\n", 20);
		exit(077);
	}
	return p;
}

static void
invalid(int c, const char *s)
{
	fprintf(stderr, "INVALID OPTION (-%c%s) - PROCESSING TERMINATED\n",
			c, s);
	exit(2);
}

#if defined (SUS) || defined (SU3) || defined (S42)
static int
regemap(int i)
{
	switch (i) {
	case REG_ESUBREG:
		return 25;
	case REG_EBRACK:
		return 49;
	case REG_EPAREN:
		return 42;
	case REG_BADBR:
	case REG_EBRACE:
		return 45;
	case REG_ERANGE:
		return 11;
	case REG_ESPACE:
#ifdef	REG_NOPAT
	case REG_NOPAT:
#endif	/* REG_NOPAT */
		return 50;
#ifdef	REG_EESCAPE
	case REG_EESCAPE:
		return 36;
#endif	/* REG_EESCAPE */
#ifdef	REG_ILLSEQ
	case REG_ILLSEQ:
		return 67;
#endif	/* REG_ILLSEQ */
	default:
		return i + 100;
	}
}
#endif	/* SUS || SU3 || S42 */

static void
settype(struct type *tp, const char *s)
{
	int	i;
#if defined (SUS) || defined (SU3) || defined (S42)
	int	reflags;
#endif

	switch (s[0]) {
	case 'p':
#if defined (SUS) || defined (SU3) || defined (S42)
		reflags = REG_ANGLES;
#if defined (SU3)
		reflags |= REG_AVOIDNULL;
#endif	/* SU3 */
		if ((i=regcomp(&tp->t_re, &s[1], reflags))!=0) {
			i = regemap(i);
#else	/* !SUS, !SU3, !S42 */
		if ((tp->t_ep = compile(&s[1], 0, 0)) == 0) {
			i = regerrno;
#endif	/* !SUS, !SU3, !S42 */
			fprintf(stderr, "%d This is the error code\n", i);
			fprintf(stderr, "Illegal Regular Expression\n");
			exit(2);
		}
		/*FALLTHRU*/
	case 'a':
	case 't':
	case 'n':
		tp->t_type = s[0];
		break;
	case '\0':
		tp->t_type = tp->t_def;
		break;
	default:
		fprintf(stderr, "%s: ", tp->t_str);
		invalid(tp->t_chr, s);
	}
}

static void
setdelim(int c, const char *s)
{
	const char	*os = s;
	wint_t	wc;
	int	n;

	if (next(wc, s, n), wc != '\0') {
		if (wc == WEOF)
			invalid(c, os);
		delim[0] = wc;
		s += n;
		if (next(wc, s, n), wc != '\0') {
			if (wc == WEOF)
				invalid(c, os);
			delim[1] = wc;
			s += n;
			if (next(wc, s, n), wc != '\0')
				invalid(c, os);
		}
	}
}

static void
setnum(int c, const char *s, long *lp, long def)
{
	char	*x;

	if (*s != '\0') {
		if (*s == '+' || *s == '-')
			invalid(c, s);
		*lp = strtol(s, &x, 10);
		if (*lp < 0 || *x != '\0')
			invalid(c, s);
	} else
		*lp = def;
}

static void
setformat(int c, const char *s)
{
	if (s[0] == 'r' && s[1] == 'z' && s[2] == '\0')
		format = FMT_RZ;
	else if (s[0] == 'l' && s[1] == 'n' && s[2] == '\0')
		format = FMT_LN;
	else if (s[0] == '\0' || (s[0] == 'r' && s[1] == 'n' && s[2] == '\0'))
		format = FMT_RN;
	else
		invalid(c, s);
}

static void
setsep(int c, const char *s)
{
	sep = s;
}

#if defined (SUS) || defined (SU3)
#define	arg()	(av[0][2] ? &av[0][2] : (ac-- > 0 ? (av++, (char *)av[0]) : \
					(invalid(av[0][1], ""), (char *)NULL)))
#else	/* !SUS, !SU3 */
#define	arg()	&av[0][2]
#endif

static void
scan(int ac, char **av)
{
	while (ac-- > 0) {
		if (av[0][0] == '-' && av[0][1] != '\0') {
#if defined (SUS) || defined (SU3)
		nxt:
#endif
			switch (av[0][1]) {
			case 'b':
				settype(&btype, arg());
				break;
			case 'f':
				settype(&ftype, arg());
				break;
			case 'h':
				settype(&htype, arg());
				break;
			case 'd':
				setdelim('d', arg());
				break;
			case 'i':
				setnum('i', arg(), &incr, 1);
				break;
			case 'l':
				setnum('l', arg(), &whichblank, 1);
				break;
			case 'n':
				setformat('n', arg());
				break;
			case 'p':
				pflag = 1;
				if (av[0][2]) {
#if defined (SUS) || defined (SU3)
					(av[0])++;
					goto nxt;
#else
					invalid(av[0][1], &av[0][2]);
#endif
				}
				break;
			case 's':
				setsep('s', arg());
				break;
			case 'v':
				setnum('v', arg(), &startnum, 1);
				break;
			case 'w':
				setnum('w', arg(), &width, 6);
				break;
			default:
				invalid(av[0][1], &av[0][2]);
			}
		} else if ((input = ib_open(av[0], 0)) == NULL) {
			fprintf(stderr, "CANNOT OPEN FILE %s\n", av[0]);
			exit(1);
		}
		av++;
	}
}

static size_t
determine(const char *lp, size_t l, struct type **curtp, int *blank)
{
	wint_t	wc;
	int	n, special;

	next(wc, lp, n);
	*blank = special = 0;
	if (wc == delim[0]) {
		lp += n, l -= n;
		next(wc, lp, n);
		if (wc == delim[1]) {
			lp += n, l -= n;
			next(wc, lp, n);
			if (wc == delim[0]) {
				lp += n, l -= n;
				next(wc, lp, n);
				if (wc == delim[1]) {
					lp +=n, l -= n;
					next(wc, lp, n);
					if (wc == delim[0]) {
						lp += n, l -= n;
						next(wc, lp, n);
						if (wc == delim[1]) {
							lp += n, l -= n;
							next(wc, lp, n);
							if (l==0 || wc=='\n') {
								*curtp = &htype;
								if (!pflag)
									number =
								startnum;
								special = 1;
							}
						}
					} else if (l == 0 || wc == '\n') {
						*curtp = &btype;
						special = 1;
					}
				}
			} else if (l == 0 || wc == '\n') {
				*curtp = &ftype;
				special = 1;
			}
		}
#ifdef	notdef	/* the standard seems to be fouled up here */
	} else if (mb_cur_max > 1 ? !iswgraph(wc) : !isgraph(wc)) {
		lp += n, l -= n;;
		while (l != 0) {
			next(wc, lp, n);
			if (mb_cur_max > 1 ? iswgraph(wc) : isgraph(wc))
				break;
			lp += n, l -= n;
		}
		if (l == 0)
			*blank = 1;
	}
#else
	} else if (l == 0 || wc == '\n')
		*blank = 1;
#endif
	if (*blank) {
		if (blanks < whichblank)
			blanks++;
		else
			blanks = 1;
	} else
		blanks = 0;
	return special == 0;
}

static size_t
mknum(char *buf, int size, unsigned long long n)
{
	char	*cp;

	if (size == 0)
		return 0;
	cp = buf + mknum(buf, size - 1, n / 10);
	*cp = n % 10 + '0';
	return cp - buf + 1;
}

#if defined (__GLIBC__) && defined (_IO_putc_unlocked)
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif

static void
print(int donumber, const char *lp, size_t l)
{
	char	numbuf[26], *np, *nq;
	const char	*sp;
	long	n;

	if (donumber) {
		mknum(numbuf, sizeof numbuf - 1, number);
		numbuf[sizeof numbuf - 1] = '\0';
		for (np = numbuf; *np == '0'; np++);
		if (&numbuf[sizeof numbuf - 1] - np > width)
			np = &numbuf[sizeof numbuf - 1 - width];
		n = width;
		for (nq = np; *nq; nq++)
			if (n > 0)
				n--;
		switch (format) {
		case FMT_RN:
			while (n--)
				putchar(' ');
			for (nq = np; *nq; nq++)
				putchar(*nq);
			break;
		case FMT_RZ:
			while (n--)
				putchar('0');
			for (nq = np; *nq; nq++)
				putchar(*nq);
			break;
		case FMT_LN:
			for (nq = np; *nq; nq++)
				putchar(*nq);
			while (n--)
				putchar(' ');
			break;
		}
		for (sp = sep; *sp; sp++)
			putchar(*sp);
	} else {
		n = width + 1;
		while (n--)
			putchar(' ');
	}
	while (l--) {
		putchar(*lp);
		lp++;
	}
}

static void
decide(char *lp, size_t l, struct type *curtp, int blank)
{
	int	gotcha = 0, tmp;

	switch (curtp->t_type) {
	case 'a':
		gotcha = !blank || blanks == whichblank;
		break;
	case 't':
		gotcha = !blank;
		break;
	case 'n':
		gotcha = 0;
		break;
	case 'p':
		if ((tmp = lp[l-1]) == '\n')
			lp[l-1] = '\0';
#if defined (SUS) || defined (SU3) || defined (S42)
		gotcha = regexec(&curtp->t_re, lp, 0, NULL, 0) == 0;
#else
		gotcha = step(lp, curtp->t_ep) != 0;
#endif
		lp[l-1] = tmp;
		break;
	}
	print(gotcha, lp, l);
	if (gotcha)
		number += incr;
}

static void
nl(struct iblok *ip)
{
	char	*line = NULL;
	size_t	linesize = 0, len;
	int	blank;
	struct type	*curtp = &btype;

	number = startnum;
	while ((len = ib_getlin(ip, &line, &linesize, srealloc)) != 0) {
		if (determine(line, len, &curtp, &blank))
			decide(line, len, curtp, blank);
		else
			putchar('\n');
	}
}

int
main(int argc, char **argv)
{
	setlocale(LC_COLLATE, "");
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	scan(argc - 1, &argv[1]);
	if (input == NULL)
		input = ib_alloc(0, 0);
	nl(input);
	return 0;
}
